### PR TITLE
WIP: Refactor cdrdao toc/table functions into Task and provide progress output

### LINKED ISSUE #299: read-toc progress information
the first stage of the rip is the `cdrdao read-toc` command which takes a few minutes to run here. it provides some output yet that is somewhat silenced by whipper and there's zero information something is going on.

it would be nice to show that output to the user to reassure them the process is not hung.

### PR DESCRIPTION
**Pull request by jtl999**
This PR fixes issue #299 and also refactors the cdrdao read toc function into a Task as a prerequisite for this. I've rebased my changes against the upstream/develop branch and ran a [test rip](https://gist.github.com/jtl999/037a40da978cdc3a7c7ff7646d3835a9) just now.

A question. Now that at present we aren't using fast_toc, do we really need to read the TOC and table separately, when it appears they contain the same data, except for the offset in the table. Would shave some time off ripping if it's possible to only need to do it once, but I feel such a change would be out of scope of this PR.

Fixes #299.

### CODE REVIEW

**Comment by jtl999** [comment]
Didn't know you used Flake8. I'll go take a look at implementing the changes so it passes later.

**Comment by JoeLametta** [comment]
💖 Thanks for opening your first pull request here! 💖

> Didn't know you used Flake8. I'll go take a look at implementing the changes so it passes later if needed/wanted.

I've just rebased your branch on whipper's `develop` and fixed flake8's warnings with a new commit.
I haven't reviewed the changes, maybe you could split them into several commits? (for increased clarity)

**Comment by jtl999** [comment]
From my personal feature branch (without the single commit rebase)

I'll look over each commit and provide a more detailed commit message for each commit.

```
commit 3445eb2804d840508319779c21c4acee22df18fa (HEAD -> verbose-toc, origin/verbose-toc, gitlab/verbose-toc)
Author: JTL <jtl@teamclassified.ca>
Date:   Fri Dec 7 23:25:02 2018 -0800

    Another attempt at getting preserve-toc working

commit 4ccf4093f4d1cd88a417f8d0d8dcfebb3124a69c
Author: JTL <jtl@teamclassified.ca>
Date:   Wed Dec 5 22:14:55 2018 -0800

    First attempt at getting toc path working

commit 0761111934c9830f95dffceb55091d4418881ea2
Author: JTL <jtl@teamclassified.ca>
Date:   Sun Dec 2 21:46:44 2018 -0800

    Provide progress of cdrdao toc reading to console

commit 825ca380227a985c47379f269aed7ce12d346ac9
Author: JTL <jtl@teamclassified.ca>
Date:   Sat Dec 1 21:35:51 2018 -0800

    Add cdrdao regexes to cdrdao.py

# This was a private notes text file for myself, not needed to be merged.
commit b89fde4439f92e268530a669cf89ebdd21783d75
Author: JTL <jtl@teamclassified.ca>
Date:   Sat Dec 1 17:40:53 2018 -0800

    Add notes for cdrdao regex parsing

commit 1e69a3db50cfd25831b8dada364e2c61e6ac3993
Author: JTL <jtl@teamclassified.ca>
Date:   Sat Dec 1 17:26:21 2018 -0800

    asynchronously be able to parse cdrdao output now

commit b9fe29d6df06ce5c8de2ab5c60d110ab4410862e
Author: JTL <jtl@teamclassified.ca>
Date:   Fri Nov 30 21:48:23 2018 -0800

    cdrdao is now a Task

commit 1a9e64ce6061bb705acb491ad2ee9b82a412fc01
Author: JTL <jtl@teamclassified.ca>
Date:   Fri Nov 30 19:23:06 2018 -0800

    Remove some unneeded code

commit a92145d36135e16227b5098bb7bfa11168e6bdd1
Author: JTL <jtl@teamclassified.ca>
Date:   Fri Nov 30 19:08:56 2018 -0800

    Work on inital attempt to move cdrdao to a Task
```

**Comment by jtl999** [comment]
Alright.

Redid my issue/299 branch privately with more commits and more of an explanation per commit. That being said I merged the bugfix/issue-335-drop-caching branch from upstream into it and then forgot that it's not technically merged into master yet.

What's the best course of action in your opinion?

**Comment by jtl999** [comment]
Managed to undo that merge via rebase and force pushed an update to issue/299

**Comment by jtl999** [comment]
A note: The progress for reading TOC/table isn't "exact" because it updates the progress upon cdrdao having started analyzing a track, as opposed after it's finished analyzing a track.

Just thought I'd mention that.

**Comment by JoeLametta** [comment]
I've just rebased your branch on whipper's develop, squashed some commits and fixed flake8's warnings with a separate commit.

**Comment by MerlijnWajer** [comment]
Why is it called ReadTOC_Task with a '_' now? (Just a nitpick)

**Comment by JoeLametta** [comment]
I've just rebased your branch on whipper's develop (fixing the conflicts) and restored the old task name (`ReadTOCTask`)... :wink:

**Comment by jtl999** [comment]
I see. Thanks!

File: whipper/program/cdrdao.py [hunk 1]
```diff
@@ -2,58 +2,162 @@
...
-        shutil.copy(tocfile, os.path.join(t_dirn, t_dst))
-    os.unlink(tocfile)
-    return toc
+    description = "Reading TOC"
+    toc = None
+
+    def __init__(self, device, fast_toc=False, toc_path=None):
+        """
+        Read the TOC for 'device'.
+        @param device:  block device to read TOC from
```
Comment (Freso) [inline whipper/program/cdrdao.py]: There should be a blank line before this one. See https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings :)
```suggestion

        @param device:  block device to read TOC from
```

File: whipper/program/cdrdao.py [hunk 2]
```diff
@@ -2,58 +2,162 @@
...
+        @type  device:  str
+        @param fast_toc:  If to use fast-toc cdrdao mode
+        @type  fast_toc: bool
+        @param toc_path: Where to save TOC if wanted.
+        @type  toc_path: str
+
```
Comment (Freso) [inline whipper/program/cdrdao.py]: There should not be a blank line here.
```suggestion
```

**Comment by JoeLametta** [comment]
> A question. Now that at present we aren't using fast_toc, do we really need to read the TOC and table separately, when it appears they contain the same data, except for the offset in the table. Would shave some time off ripping if it's possible to only need to do it once, but I feel such a change would be out of scope of this PR.

Related issue #173.
More information about `fast_toc` in #264. :wink:

File: whipper/program/cdrdao.py [hunk 3]
```diff
@@ -63,10 +167,7 @@ def DetectCdr(device):
     cmd = [CDRDAO, 'disk-info', '-v1', '--device', device]
     logger.debug("executing %r", cmd)
     p = Popen(cmd, stdout=PIPE, stderr=PIPE)
-    if 'CD-R medium          : n/a' in p.stdout.read():
-        return False
-    else:
-        return True
+    return 'CD-R medium          : n/a' not in p.stdout.read()
```
Comment (Freso) [inline whipper/program/cdrdao.py]: Is this related to this PR? Seems like the underlying logic of this isn't changed, just how it's written.

Comment (jtl999) [reply whipper/program/cdrdao.py]: No. @JoeLametta did that. It behaves identically in my testing with a pressed CD but is simplified.

**Comment by jtl999** [comment]
Implemented the style changes.

With regards to your comment on IRC (about the mix of single and double quotes) how do you feel it should be done? I was trying to emulate the style of whipper/programs/cdparanoia.py (Which I read to understand the task system)

File: whipper/program/cdrdao.py (hunk 3 above)
Comment (Freso) [reply whipper/program/cdrdao.py]: Right. My point is that it's a change that doesn't belong in this PR, even if the fault is not yours by @JoeLametta's. It adds on to the code that reviewers have to look through, even though it isn't related to the topic of "Refactor cdrdao toc/table functions into Task and provide progress output". It should be its own PR. (Small PRs are okay too, in fact, I'd rather review a lot of small, self-contained PRs than a bunch of big ones containing random other fixes.)

@JoeLametta?

**Comment by Freso** [comment]
> With regards to your comment on IRC (about the mix of single and double quotes) how do you feel it should be done?

I usually go for single quoting in my own projects, but am fine with double quotes too. Just as long as it's consistent. Anyway, I didn't remark on it on my previous comment to this PR, because it is irrelevant to this PR. Please just keep your commits [atomic](https://www.freshconsulting.com/atomic-commits/) and your PRs self-contained. If we want to fix quotation style, that is its own PR (or something that can be done gradually in other PRs once it's been decided on one way or another).

File: whipper/program/cdrdao.py [hunk 4]
```diff
@@ -2,71 +2,172 @@
...
+        self.fd, self.tocfile = tempfile.mkstemp(
+            suffix=u'.cdrdao.read-toc.whipper.task')
+
+    def start(self, runner):
+        task.Task.start(self, runner)
+        os.close(self.fd)
+        os.unlink(self.tocfile)
+
+        cmd = ([CDRDAO, 'read-toc'] + (['--fast-toc'] if self.fast_toc
+               else []) + ['--device', self.device, self.tocfile])
```
Comment (Freso) [inline whipper/program/cdrdao.py]: Another stylistic change. I'd probably do something like
```python
        cmd = ([CDRDAO, 'read-toc']
               + (['--fast-toc'] if self.fast_toc else [])
               + ['--device', self.device, self.tocfile])
```
or even
```python
        cmd = (
            [CDRDAO, 'read-toc']
            + (['--fast-toc'] if self.fast_toc else [])
            + ['--device', self.device, self.tocfile]
        )
```
To make the sections easier to recognise. Putting a linebreak in the middle of a sub command seems ugly. (Breaking before binary operators as per [PEP 8: Should a Line Break Before or After a Binary Operator?](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator))

**Comment by JoeLametta** [comment]
@Freso Sorry for the mess, I've redone everything from scratch. Hope it's better this time...

File: whipper/program/cdrdao.py (hunk 3 above)
Comment (JoeLametta) [reply whipper/program/cdrdao.py]: Sorry, I've reverted this change.

**Comment by jtl999** [comment]
> can we please try and keep PRs self-contained?

I agree.

I will review this

**Comment by JoeLametta** [comment]
@Freso Do you think latest revision of this pull requests is ready for merging?

**Comment by jtl999** [comment]
Something that might be out of scope for this PR but worthwhile for the 1.0 milestone: My uncle pointed out recently when I was showing him the code and explaining to him how it works (He's been a long time Linux user, but not much of a programmer outside of very trivial Python tasks and specifically was an early user of cdparanoia/cdda2wav and EAC in late 90's.

He thinks the functionality of the code is great but it should have some documentation explaining the concept of Q sub-channel errors and what it means for ripping the disc. In my case with a TSSTCorp SH-S223L and some scratched CD's I tested with, getting above 1500 Q sub-channel errors on a track usually means I cant reliably rip said track because whipper gets CRC mismatches in Test and Copy mode.

**Comment by JoeLametta** [comment]
> Something that might be out of scope for this PR but worthwhile for the 1.0 milestone

:+1:

**Comment by Freso** [comment]
This introduced https://github.com/whipper-team/whipper/issues/374
